src/Microsoft.ML.Data/Scorers/FeatureContributionCalculation.cs

```diff
@@ -0,0 +1,451 @@
+            // bool: stringify
+
+            ctx.SaveModel(Predictor, ModelFileUtils.DirPredictor);
+            Contracts.Assert(0 <= _topContributionsCount);
+            ctx.Writer.Write(_topContributionsCount);
+            Contracts.Assert(0 <= _bottomContributionsCount);
+            ctx.Writer.Write(_bottomContributionsCount);
+            ctx.Writer.WriteBoolByte(_normalize);
+            ctx.Writer.WriteBoolByte(Stringify);
+        }
+
+        public Delegate GetTextContributionGetter(Row input, int colSrc, VBuffer<ReadOnlyMemory<char>> slotNames)
+        {
+            Contracts.CheckValue(input, nameof(input));
+            Contracts.Check(0 <= colSrc && colSrc < input.Schema.Count);
+            var typeSrc = input.Schema[colSrc].Type;
+
+            Func<Row, int, VBuffer<ReadOnlyMemory<char>>, ValueGetter<ReadOnlyMemory<char>>> del = GetTextValueGetter<int>;
+            var meth = del.GetMethodInfo().GetGenericMethodDefinition().MakeGenericMethod(typeSrc.RawType);
+            return (Delegate)meth.Invoke(this, new object[] { input, colSrc, slotNames });
+        }
+
+        public Delegate GetContributionGetter(Row input, int colSrc)
+        {
+            Contracts.CheckValue(input, nameof(input));
+            Contracts.Check(0 <= colSrc && colSrc < input.Schema.Count);
+
+            var typeSrc = input.Schema[colSrc].Type;
+            Func<Row, int, ValueGetter<VBuffer<float>>> del = GetValueGetter<int>;
+
+            // REVIEW: Assuming Feature contributions will be VBuffer<float>.
+            // For multiclass LR it needs to be(VBuffer<float>[].
+            var meth = del.GetMethodInfo().GetGenericMethodDefinition().MakeGenericMethod(typeSrc.RawType);
+            return (Delegate)meth.Invoke(this, new object[] { input, colSrc });
+        }
+
+        private ReadOnlyMemory<char> GetSlotName(int index, VBuffer<ReadOnlyMemory<char>> slotNames)
+        {
+            _env.Assert(0 <= index && index < slotNames.Length);
+            var slotName = slotNames.GetItemOrDefault(index);
+            return slotName.IsEmpty
+                ? new ReadOnlyMemory<char>($"f{index}".ToCharArray())
+                : slotName;
+        }
+
+        private ValueGetter<ReadOnlyMemory<char>> GetTextValueGetter<TSrc>(Row input, int colSrc, VBuffer<ReadOnlyMemory<char>> slotNames)
+        {
+            Contracts.AssertValue(input);
+            Contracts.AssertValue(Predictor);
+
+            var featureGetter = input.GetGetter<TSrc>(colSrc);
+            var map = Predictor.GetFeatureContributionMapper<TSrc, VBuffer<float>>(_topContributionsCount, _bottomContributionsCount, _normalize);
+
+            var features = default(TSrc);
+            var contributions = default(VBuffer<float>);
+            return
+                (ref ReadOnlyMemory<char> dst) =>
+                {
+                    featureGetter(ref features);
+                    map(in features, ref contributions);
+                    int[] indices;
```

Comment:
>indices [](start = 26, length = 7)

Instead of allocating new arrays, you can write it this way:

var editor = VBufferEditor.CreateFromBuffer(ref contributions);
                        var indices = contributions.IsDense ? Utils.GetIdentityPermutation(contributions.Length) : editor.Indices;
                        var values = editor.Values;

Reply:
Thank you that looks a lot better now!

---
In reply to: [242993437](https://github.com/dotnet/machinelearning/pull/1847#discussion_r242993437) [](ancestors = 242993437)